### PR TITLE
DTPAYETWO-485: Proxy Implementation

### PR DESCRIPTION
**Pull request by hlahlou-pp-dev**
Ticket: https://engineering.paypalcorp.com/jira/browse/DTPAYETWO-485

- Modified HTTP package and supplied it as a local util
- Modified Hyperwallet  and apiClient obj to take in proxy configuration as well as all HTTP methods
- Created relevant Functional and Unit Tests

Testing:
Used proxy and monitored status  of incoming requests and responses

Screen Shot 2022-02-14 at 2 20 39 PM
Screen Shot 2022-02-14 at 2 20 11 PM
Screen Shot 2022-02-14 at 2 14 23 PM

### CODE REVIEW

File: src/main/java/com/hyperwallet/clientsdk/util/HyperwalletApiClient.java [hunk 1]
```diff
@@ -169,8 +175,13 @@ private String getAuthorizationHeader() {
 
     private Request getService(final String url, boolean isHttpGet) {
         String contentType = "application/" + ((isEncrypted) ? "jose+json" : "json");
-        Request request = new Request(url)
-                .addHeader("Authorization", getAuthorizationHeader())
+        Request request;
```
Comment (skoong) [inline src/main/java/com/hyperwallet/clientsdk/util/HyperwalletApiClient.java:174]: Minor: May be cleaner to use ternary operator here
```
Request request = usesProxy() ? new Request(url, proxy) : new Request(url);
```

File: src/main/java/com/hyperwallet/clientsdk/util/HyperwalletApiClient.java [hunk 2]
```diff
@@ -154,6 +157,9 @@ protected void checkErrorResponse(final Response response) throws ParseException
     private void checkResponseHeader(Response response) {
         String contentTypeHeader = response.getHeader(CONTENT_TYPE_HEADER);
         String expectedContentType = isEncrypted ? VALID_JSON_JOSE_CONTENT_TYPE : VALID_JSON_CONTENT_TYPE;
+//        if (response.getResponseCode() == 302) {
```
Comment (skoong) [inline src/main/java/com/hyperwallet/clientsdk/util/HyperwalletApiClient.java]: Do we need this commented out code?

Comment (peter-joseph) [inline src/main/java/com/hyperwallet/clientsdk/util/HyperwalletApiClient.java:160]: please remove commented code

File: src/main/java/com/hyperwallet/clientsdk/util/Message.java [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+ */
+
+/**
+ * This class represents an HTTP Message, which could either be a Request or a Response. Message is an abstract class that contains fields and methods that are common to both types of Messages.
+ *
+ * @param <T>
+ *            A Type that extends Message (either {@link Request} or {@link Response})
+ */
+public abstract class Message<T extends Message<T>> {
```
Comment (peter-joseph) [inline src/main/java/com/hyperwallet/clientsdk/util/Message.java:39]: is this copied code?, can you make sure that we are not legally bounded with this approach? or can we also add something to recognize this persons work, please work with legal thanks.

File: src/main/java/com/hyperwallet/clientsdk/util/Request.java [hunk 4]
```diff
@@ -0,0 +1,436 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * This class represents an HTTP Request message.
+ */
+public class Request extends Message<Request> {
```
Comment (peter-joseph) [inline src/main/java/com/hyperwallet/clientsdk/util/Request.java:48]: this code too please make sure we are not legally bounded on this, thanks

**Review by peter-joseph — COMMENTED** [review]
Looks good but please make sure that we have rights on those copied codes, just confirm it in this PR and this should be good.

File: src/main/java/com/hyperwallet/clientsdk/util/HyperwalletApiClient.java (hunk 2 above)
Comment (hlahlou-pp-dev) [reply src/main/java/com/hyperwallet/clientsdk/util/HyperwalletApiClient.java]: This is outdated, I cleaned up the comments but hadn't pushed the change

File: src/main/java/com/hyperwallet/clientsdk/util/HyperwalletApiClient.java (hunk 1 above)
Comment (hlahlou-pp-dev) [reply src/main/java/com/hyperwallet/clientsdk/util/HyperwalletApiClient.java:174]: I do think this is a cleaner change, will impl

**Review by skoong — APPROVED** [review]

File: src/main/java/com/hyperwallet/clientsdk/util/Request.java (hunk 4 above)
Comment (hlahlou-pp-dev) [reply src/main/java/com/hyperwallet/clientsdk/util/Request.java:48]: Screen Shot 2022-02-15 at 9 52 31 AM

We should be covered for this

**Review by peter-joseph — APPROVED** [review]

**Pull request merged**